pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -517,40 +620,77 @@ class StateVectorKokkos final
     }
 
+    /**
+     * @brief Apply a given matrix for controlled operations directly to the
+     * statevector using a raw matrix pointer vector.
+     *
+     * @param matrix Pointer to the array data (in row-major format).
+     * @param controlled_wires Controlled wires
+     * @param controlled_values Controlled values (true or false)
+     * @param wires Wires to apply gate to.
+     * @param inverse Indicate whether inverse should be taken. (Default to
+     * false)
+     */
+    inline void applyControlledMatrix(
+        ComplexT *matrix, const std::vector<std::size_t> &controlled_wires,
+        const std::vector<bool> &controlled_values,
+        const std::vector<std::size_t> &wires, bool inverse = false) {
+        PL_ABORT_IF(wires.empty(), "Number of wires must be larger than 0");
+        const std::size_t n2 = exp2(wires.size() * 2);
+        KokkosVector matrix_("matrix_", n2);
+        Kokkos::deep_copy(matrix_, UnmanagedComplexHostView(matrix, n2));
+        applyNCMultiQubitOp(matrix_, controlled_wires, controlled_values, wires,
+                            inverse);
+    }
+
     /**
      * @brief Apply a given matrix directly to the statevector using a
      * raw matrix pointer vector.
      *
      * @param matrix Pointer to the array data (in row-major format).
+     * @param controlled_wires Controlled wires
+     * @param controlled_values Controlled values (true or false)
      * @param wires Wires to apply gate to.
-     * @param inverse Indicate whether inverse should be taken.
-     */
-    inline void applyMatrix(const ComplexT *matrix,
-                            const std::vector<std::size_t> &wires,
-                            bool inverse = false) {
+     * @param inverse Indicate whether inverse should be taken. (Default to
+     * false)
+     */
+    inline void
+    applyControlledMatrix(const ComplexT *matrix,
+                          const std::vector<std::size_t> &controlled_wires,
+                          const std::vector<bool> &controlled_values,
+                          const std::vector<std::size_t> &wires,
+                          bool inverse = false) {
         PL_ABORT_IF(wires.empty(), "Number of wires must be larger than 0");
-        std::size_t n = static_cast<std::size_t>(1U) << wires.size();
-        std::size_t n2 = n * n;
+        const std::size_t n2 = exp2(wires.size() * 2);
         KokkosVector matrix_("matrix_", n2);
         Kokkos::deep_copy(matrix_, UnmanagedConstComplexHostView(matrix, n2));
-        applyMultiQubitOp(matrix_, wires, inverse);
+        applyNCMultiQubitOp(matrix_, controlled_wires, controlled_values, wires,
+                            inverse);
     }
 
     /**
-     * @brief Apply a given matrix directly to the statevector.
+     * @brief Apply a given controlled-matrix directly to the statevector.
      *
-     * @param matrix Matrix data (in row-major format).
+     * @param matrix Vector containing the statevector data (in row-major
```

Comment:
is the `matrix` only for the target wires?

Reply:
Yes, it's only for target wires.